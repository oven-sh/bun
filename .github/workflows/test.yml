name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
