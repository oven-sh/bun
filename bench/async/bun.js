import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
