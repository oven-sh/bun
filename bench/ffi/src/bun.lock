{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
