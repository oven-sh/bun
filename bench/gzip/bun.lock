{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
