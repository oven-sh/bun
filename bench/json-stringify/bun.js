import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
