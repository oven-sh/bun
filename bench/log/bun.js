import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
