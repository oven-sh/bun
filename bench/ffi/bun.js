import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
