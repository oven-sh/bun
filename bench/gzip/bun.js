import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
