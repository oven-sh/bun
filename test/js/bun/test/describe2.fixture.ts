import { describe, test, expect, beforeAll, beforeEach, afterEach, afterAll } from "bun:test";

console.log("enter");

describe("describe 1", () => {
  console.log("describe 1");
  describe("describe 2", () => {
    console.log("describe 2");
  });
  describe("describe 3", () => {
    console.log("describe 3");
  });
});
describe("describe 4", () => {
  console.log("describe 4");
  describe("describe 5", () => {
    console.log("describe 5");
    describe("describe 6", () => {
      console.log("describe 6");
    });
    describe("describe 7", () => {
      console.log("describe 7");
    });
  });
});
describe("describe 8", () => {
  console.log("describe 8");
});
describe.each([1, 2, 3, 4])("describe each %s", i => {
  console.log(`describe each ${i}`);
  describe.each(["a", "b", "c", "d"])("describe each %s", j => {
    console.log(`describe each ${i}${j}`);
  });
});

describe("failed describe", () => {
  console.log("failed describe");
  test("in failed describe", () => {
    console.log("this test should not run because it is in a failed describe");
  });
  describe("failed describe inner 1", () => {
    console.log("failed describe inner 1");
    test("in failed describe inner 1", () => {
      console.log("this test should not run because it is in a failed describe inner 1");
    });
  });
  describe("failed describe inner 2", () => {
    console.log("failed describe inner 2");
  });
  throw "uh oh";
});

// == async ==

describe("async describe 1", async () => {
  console.log("async describe 1");
  describe("async describe 2", async () => {
    console.log("async describe 2");
  });
  describe("async describe 3", async () => {
    console.log("async describe 3");
  });
});
describe("async describe 4", async () => {
  console.log("async describe 4");
  describe("async describe 5", async () => {
    console.log("async describe 5");
  });
  describe("async describe 6", async () => {
    console.log("async describe 6");
  });
});

describe("actual tests", () => {
  test("more functions called after delayed done", done => {
    process.nextTick(() => {
      done();
      throw "uh oh";
    });
  });
  test("another test", async () => {
    expect(true).toBe(true);
  });
});

// == concurrent ==

describe.concurrent("concurrent describe 1", () => {
  test("item 1", async () => {});
  test("item 2", async () => {});
  test.failing("snapshot in concurrent group", async () => {
    console.log("snapshot in concurrent group");
    // this is a technical limitation of not using async context. in the future, we could allow thisa
    expect("hello").toMatchSnapshot();
  });
});

// == other stuff ==

test("LINE 66", () => console.log("LINE 66"));
test.skip("LINE 67", () => console.log("LINE 67"));
test.failing("LINE 68", () => console.log("LINE 68"));
test.todo("LINE 69", () => console.log("LINE 69"));
test.each([1, 2, 3])("LINE 70", item => console.log("LINE 70", item));
test.if(true)("LINE 71", () => console.log("LINE 71"));
test.skipIf(true)("LINE 72", () => console.log("LINE 72"));
test.concurrent("LINE 74", () => console.log("LINE 74"));
test.todo("failing todo passes", () => {
  throw "this error is shown";
});
test.failing("failing failing passes", () => {
  throw "this error is not shown";
});

// == timeout ==
test("this test times out", () => Bun.sleep(100), 1);
test("this test times out with done", done => {}, 1);

// == each ==
test.each([
  [1, 2, 3],
  [2, 3, 5],
  [3, 4, 7],
])("addition %i + %i = %i", (a, b, expected) => {
  console.log(`adding: ${a} + ${b} = ${expected}`);
  expect(a + b).toBe(expected);
});

// == expect.assertions/hasAssertions ==
test.failing("expect.assertions", () => {
  // this test should fail despite being 'test.failing', matching existing behaviour
  // we might consider changing this.
  expect.assertions(1);
  expect.hasAssertions(); // make sure this doesn't overwrite the assertions count, matching existing behaviour
});

test.concurrent.failing("expect.assertions not yet supported in concurrent tests", () => {
  expect.hasAssertions(); // this call will fail because expect.hasAssertions is not yet supported in concurrent tests
  expect(true).toBe(true);
});
test.concurrent.failing("expect.assertions not yet supported in concurrent tests", () => {
  expect.assertions(1); // this call will fail because expect.assertions is not yet supported in concurrent tests
  expect(true).toBe(true);
});

test("expect.assertions works", () => {
  expect.assertions(2);
  expect(true).toBe(true);
  expect(true).toBe(true);
});

// === timing edge case ===
test.failing("more functions called after delayed done", done => {
  process.nextTick(() => {
    done();
    expect(true).toBe(false);
  });
});
test("another test", async () => {});

// === timing failure case. if this is fixed in the future, update the test ===
test("misattributed error", () => {
  setTimeout(() => {
    expect(true).toBe(false);
  }, 10);
});
test.failing("passes because it catches the misattributed error", done => {
  setTimeout(done, 50);
});

// === hooks ===
describe("hooks", () => {
  beforeAll(() => {
    console.log("beforeAll1");
  });
  beforeEach(async () => {
    console.log("beforeEach1");
  });
  afterAll(done => {
    console.log("afterAll1");
    done();
  });
  afterEach(done => {
    console.log("afterEach1");
    Promise.resolve().then(done);
  });
  afterEach(() => {
    console.log("afterEach2");
  });
  afterAll(() => {
    console.log("afterAll2");
  });
  beforeAll(async () => {
    console.log("beforeAll2");
  });
  beforeEach(() => {
    console.log("beforeEach2");
  });
  test("test1", () => {
    console.log("test1");
  });
  test("test2", () => {
    console.log("test2");
  });
});

// === done parameter ===
describe("done parameter", () => {
  test("instant done", done => {
    done();
  });
  test("delayed done", done => {
    setTimeout(() => {
      done();
    }, 1);
  });
  describe("done combined with promise", () => {
    let completion = 0;
    beforeEach(() => (completion = 0));
    afterEach(() => {
      if (completion != 2) throw "completion is not 2";
    });
    test("done combined with promise, promise resolves first", async done => {
      setTimeout(() => {
        completion += 1;
        done();
      }, 200);
      await Bun.sleep(50);
      completion += 1;
    });
    test("done combined with promise, done resolves first", async done => {
      setTimeout(() => {
        completion += 1;
        done();
      }, 50);
      await Bun.sleep(200);
      completion += 1;
    });
    test("fails when completion is not incremented", () => {});
  });
  describe("done combined with promise error conditions", () => {
    test.failing("both error and done resolves first", async done => {
      done("test error");
      throw "promise error";
    });
    test.failing("both error and promise resolves first", async done => {
      setTimeout(() => done("done error"), 10);
      throw "promise error";
    });
    test.failing("done errors only", async done => {
      done("done error");
    });
    test.failing("promise errors only", async done => {
      setTimeout(() => done(), 10);
      throw "promise error";
    });
  });
  test("second call of done callback ignores triggers error", done => {
    done();
    done("uh oh!");
  });
});

test.failing("microtasks and rejections are drained after the test callback is executed", () => {
  Promise.reject(new Error("uh oh!"));
});

console.log("exit");
