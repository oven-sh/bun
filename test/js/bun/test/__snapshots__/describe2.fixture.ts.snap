// Bun Snapshot v1, https://bun.sh/docs/test/snapshots
